### ISSUE TITLE
Question in tf.nn.sigmoid_cross_entropy_with_logits(labels=y, logits=logits)

### ISSUE THREAD

**Issue by JeffreyYafeiHu** [ageron/handson-ml#121]
Hi Aurélien, I'd like to ask why in `tf.nn.sigmoid_cross_entropy_with_logits(labels=y, logits=logits)`
the logits is logits( tf.layers.dense(hidden, n_outputs)) instead of outputs (tf.nn.sigmoid(logits))

Following are relevant code,
```
logits = tf.layers.dense(hidden, n_outputs)
outputs = tf.nn.sigmoid(logits)  # probability of action 0 (left)
p_left_and_right = tf.concat(axis=1, values=[outputs, 1 - outputs])
action = tf.multinomial(tf.log(p_left_and_right), num_samples=1)

y = 1. - tf.to_float(action)
cross_entropy = tf.nn.sigmoid_cross_entropy_with_logits(labels=y, logits=logits)
```

**Comment by ageron** [comment]
Hi @JeffreyYafeiHu ,
Thanks for your question.  Indeed, this confused me as well at first.  I wrote a note that answers your question on page 268 of the book:

> The `sparse_softmax_cross_entropy_with_logits()` function is equivalent to applying the softmax activation function and then computing the cross entropy, but it is more efficient, and it properly takes care of corner cases: when logits are large, floating-point rounding errors may cause the softmax output to be exactly equal to 0 or 1, and in this case the cross entropy equation would contain a log(0) term, equal to negative infinity. The `sparse_softmax_cross_entropy_with_logits()` function solves this problem by computing log(ε) instead, where ε is a tiny positive number. This is why we did not apply the softmax activation function earlier. There is also another function called `softmax_cross_entropy_with_logits()`, which takes labels in the form of one-hot vectors (instead of ints from 0 to the number of classes minus 1).

Hope this helps.

**Comment by JeffreyYafeiHu** [comment]
@ageron Thank you for the explanation!
